src/core/processing/qgsprocessingregistry.cpp

```diff
@@ -125,3 +153,26 @@ QgsProcessingAlgorithm *QgsProcessingRegistry::createAlgorithmById( const QStrin
   return creation.release();
 }
 
+void QgsProcessingRegistry::addParameterType( QgsProcessingParameterType *type )
+{
+  mParameterTypes.insert( type->id(), type );
```

Comment:
I think here we should first check whether an existing type with the same id() exists, and if so, throw an error to the message log. It's possible plugins may accidentally duplicate each other's ids. Then this method could return a bool to indicate success/failure of the add.

(What I've never been sure on is how to handle the ownership if something like this fails - I usually just delete the object immediately, since ownership WAS transferred and it prevents a leak. But I'm not sure if that's good practice or not...)

Reply:
Maybe SIP_TRANSFER and SIP_TRANSFERBACK where you return a nullptr in case it's accepted?

Reply:
That sounds like a great recipe for wasting a good half day fighting with random sip behaviour :p

Reply:
are you sure of that?

Reply:
100%. My big take away from 3.0 is NEVER to try anything fancy with sip, and always go for the simple c++ only solutions.